presto-hive/src/main/java/com/facebook/presto/hive/HiveType.java

```diff
@@ -225,7 +228,14 @@ private static TypeInfo toTypeInfo(Type type)
             return HIVE_DOUBLE.typeInfo;
         }
         if (type instanceof VarcharType) {
-            return HIVE_STRING.typeInfo;
+            VarcharType varcharType = (VarcharType) type;
+            int varcharLength = varcharType.getLength();
+            if (varcharLength <= MAX_VARCHAR_LENGTH) {
+                return getVarcharTypeInfo(varcharLength);
+            }
+            else if (varcharLength == MAX_LENGTH) {
+                return HIVE_STRING.typeInfo;
+            }
```

Comment:
It would be nice to throw specific exception here so user knows what values for precision parameter are supported.